wallet/src/state.rs

```diff
@@ -188,12 +227,168 @@ impl ClientState {
+                amount: rate_data.delegation_amount(unbonded_amount),
+                asset_id: rate_data.identity_key.delegation_token().id(),
+            },
+            memo::MemoPlaintext([0u8; memo::MEMO_LEN_BYTES]),
+            self.wallet.outgoing_viewing_key(),
+        );
+
+        let change_amount = spent_amount - spend_amount;
+        // TODO: support dummy notes, and produce a change output unconditionally.
+        // let change_note = if change_amount > 0 { ... } else { /* dummy note */}
+        if change_amount > 0 {
+            let change_note = tx_builder.add_output_producing_note(
+                rng,
+                &self_address,
+                Value {
+                    amount: change_amount,
+                    asset_id: *STAKING_TOKEN_ASSET_ID,
+                },
+                memo::MemoPlaintext([0u8; memo::MEMO_LEN_BYTES]),
+                self.wallet.outgoing_viewing_key(),
+            );
+            self.register_change(change_note);
+        }
+
+        self.register_change(delegation_note);
+
+        tx_builder.finalize(rng).map_err(Into::into)
+    }
+
+    /// Generate a new transaction delegating stake
+    #[instrument(skip(self, rng))]
+    pub fn build_undelegate<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        rate_data: RateData,
+        delegation_amount: u64,
+        fee: u64,
+        source_address: Option<u64>,
+    ) -> Result<Transaction, anyhow::Error> {
+        // If the source address is set, send the delegation tokens to the same
+        // address; otherwise, send them to the default address.
+        let (_label, self_address) = self
+            .wallet()
+            .address_by_index(source_address.unwrap_or(0) as usize)?;
+
+        let mut tx_builder = Transaction::build_with_root(self.note_commitment_tree.root2());
+
+        tx_builder
+            .set_fee(fee)
+            // XXX populate chain id from params?
+            .set_chain_id(CURRENT_CHAIN_ID.to_string())
+            .add_undelegation(&rate_data, delegation_amount);
+
+        // Because the outputs of an undelegation are quarantined, we want to
+        // avoid any unnecessary change outputs, so we pay fees out of the
+        // unbonded amount.
+        let unbonded_amount = rate_data.unbonded_amount(delegation_amount);
+        let output_amount = unbonded_amount.checked_sub(fee).ok_or_else(|| {
+            anyhow::anyhow!(
+                "unbonded amount {} from delegation amount {} is insufficient to pay fees {}",
+                unbonded_amount,
+                delegation_amount,
+                fee
+            )
+        })?;
+
+        let delegation_denom = rate_data.identity_key.delegation_token().denom();
+
+        // XXX if the undelegation is for less than their total amount of delegation tokens,
+        // all of their remaining delegation tokens will also be quarantined.
+        // this sucks lmao
+        let mut spent_amount = 0;
```

Comment:
Per discussion on discord, the solution to this is to pre-prepare a note with the exact number of delegation tokens.  Doing this will require some other changes to be able to track transaction completion.